keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -0,0 +1,276 @@
+And those "wild" request doesn't match any buckets will be rejected.
+
+##### Overview
+
+Within each priority band, we have a set of buckets matching the priority and
+a WRR request queue. Request that missed reserved quota will be queued into
+the WRR queue w/ the weight of its bucket. Note that reserving quota of
+buckets at priority N can also re-claim requests queued at N.
+
+All requests will be pre-processed before it's actually executed by 3 stages:
+
+0. We find a matched bucket by applying BucketBinding API.
+1. *Reserving*: We check if the bucket's reserved quota. If it's sufficient,
+the request will consume one reserved quota from the bucket and be executed
+immediately.
+2. *Prioritizing*: We classify and enqueue the request into a priority band.
+Requests at higher priority will __always__ be executed prior to requests at lower.
+3. *Distributing*: We poll requests from each priority band's queue by
+[Weighted Round Robin](https://en.wikipedia.org/wiki/Weighted_round_robin)
+algorithm. If the request's matched bucket's weight is higher, its request
+will have more chance to be executed.
+
+Each stage works separately and solves separate goals.
+
+```
+                    +-----------+  enqueued  +--------------+  pending  +--------------+ polled by WRR
+Inbound Request --> | Reserving | -------->  | Prioritizing | --------> | Distributing |  ------->  Executed
+                    +-----------+            +--------------+           +--------------+
+              sufficient |             insufficient |                          |
+          reserved quota |        STSB shared quota |                          |
+                         v                          v                          v
+                 Executed Immediately           Rejected                    Time-out
+```
+
+
+##### API Model and APIServer flags
+
+
+
+```go
+type Bucket struct {
+    ...
+
+    // ReservedQuota defines a max concurrency of requests can run under guarantee w/o competing.
+    ReservedQuota int
+
+    // SharedQuota defines a shared concurrency added to the system's shared bandwidth.
+    SharedQuota int
+
+    // Priority assigns the bucket into a certain priority band.
+    Priority PriorityBand // Ranged in 1...N
+
+    // Weight defines how much weight its matched requests have in the queue.
+    Weight int
+}
+
+type BucketBinding struct {
+    ...
+
+    // Rules defines how we bind request to a bucket.
+    Rules []BucketBindingRule // WIP
+
+    // BucketRef references a bucket.
+    BucketRef *Bucket
+}
+```
+
+Also, we will add two optional flags to apiserver:
+
+- *--initial-extra-shared-quota*: default to 200-ish. Extra shared quota will be consumed when a request fails to get any reserved or shared quota from the defined buckets. Or we can say it's a "lowest" priority which is capable of running requests from any priority band. All requests queued in this "lowest" priority will be weighted evenly.
```

Comment:
Also, I don't think it's a good idea to have absolute numbers (like 200) as command line flag values. I'd prefer to see a percentage, e.g., save half of the available quota for the shared pool.

Reply:
won't this make the so-called "special bucket" non-trival and more complex? i'm slightly leaning on a most simple approach at the first stage. we could auto-tune this "extra shared quota" by the percentage as an enhancement, wdyt?

Reply:
I really dislike flags :) The complexity is there either way, it's just a question of where we put it / what entities are allowed to change it / when it can change.

I think it's reasonable to change this dynamically at runtime?

Another reason for it to be expressed in percent is that it can then be the same across apiservers, even if they are on differently sized machines.

I think this parameter is really about how much quota is reserved for unimportant users, and as such, it's a policy choice and not something we can deduce from the environment.

Reply:
oh well, initially, i thought the flag is an easier way to start in the first stage. as i commented above, wdyt if we make it a default bucket which is  ensured during the apiserver's launching. user might read/modify it at runtime. otherwise, we can make users only read it but any modification on it won't lead to any reloading (just like how we do in RBAC node rules)